docs/source/user_guide/configuration.rst
Configuration
=============

Snr can not only be configured at runtime using flags, but also using environment variables and the config file.

Here, you will find a list of configuration keys and some help on them.

The configuration file can be found in `~/.config/snr/main.conf`.
We use the INI format, all snr configuration are inside the `[main]` section.

Configuration Priority
----------------------

If configuration can come from 3 sources, which have priority over the other?

1. command line options 
2. environment variables
3. configuration file

Meaning, command line options can override both environment variables and configuration in the file, and neither of those can override command line options.

Configuration Keys
------------------

verbose
^^^^^^^

Controls the verbosity.

**Type: boolean**

**Default: off**

**Env var: SNR_VERBOSE**

quiet
^^^^^

Controls whatever informational messages are shown or not. Also suppresses the initial banner.

**Note: Setting both quiet and verbose as on will cancel each other out**

**Type: boolean**

**Default: off**

**Env var: SNR_QUIET**

arch
^^^^

Architecture of the generated image. Either `i386` or `x86_64`.

**Type: string**

**Default: Determined on runtime**

**Env var: SNR_ARCH**


default_exit_code
^^^^^^^^^^^^^^^^^

Default exit code of snr when the user runs `exit`, also used by other functions to exit normally.

**Type: integer**

**Default: 0**

**Env var: SNR_DEFAULT_EXIT_CODE**


default_hostname
^^^^^^^^^^^^^^^^

Sets the default hostname of the generated host OS.

**Type: string**

**Default: "snr"**

**Env var: SNR_HOST_HOSTNAME**

default_primary_nameserver
^^^^^^^^^^^^^^^^^^^^^^^^^^

Sets the primary nameserver (primary DNS server) of the generated host OS.

**Type: IP address (v4 only)**

**Default: "1.1.1.1"**

**Env var: SNR_HOST_PRIMARY_NAMESERVER**

default_secondary_nameserver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Sets the secondary nameserver (secondary DNS server) of the generated host OS.

**Type: IP address (v4 only)**

**Default: "1.0.0.1"**

**Env var: SNR_HOST_PRIMARY_NAMESERVER**

default_user_agent
^^^^^^^^^^^^^^^^^^

Sets the default user agent for download operations.

**Type: string**

**Default: "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/113.0"**

**Env var: SNR_USER_AGENT**

.. versionadded:: 1.1.0


CHANGELOG
# Changelog

## Quick Access

- [Legacy Versions](#legacy-versions)
- [Stable](#v15x-stable)
- [Beta](#v16x)
- [v1.0.X](#v10x)
- [v1.1.X](#v11x)
- [v1.2.X](#v12x)
- [v1.3.X](#v13x)
- [v1.4.X](#v14x)
- [v1.5.X](#v15x-stable)
- [v1.6.X](#v16x)

## Legacy Versions

### v0.0.1

Initial release.

### v0.1.0

- Additions:

  - `info` command.

  - Ensure traceback is distinct enough.

  - Add payload count to the banner.

- Bugfixes:

  - Fix banner coloring issues.

Some under-the-hood changes and code improvements.

## v1.0.X

### v1.0.0

This release is a huge step forward for snr, truly its transformation into an application worthy.

- Additions:

  - editorconfig file to ensure style across IDEs and text editors.

  - Now using logging methods automatically writes them to a log file as well.

  - Use docker and docker compose.

  - Payload class, with most common functionality needed for payloads as functions to ease writing payloads even more.

  - New utilities (like `bytes_to_str_repr`.)

  - New `ProgramWrapper` options: `fakeroot`, `fakechroot` and `sudo`. It allows you to run a program with fake root permissions, prepare for a fakechroot environment and ask for root using `sudo`.

  - Now program wrapper's expose the `SubprocessError` exception class for better access.

  - A huge set of program wrappers have been added.

  - Add builtin debugger support (for when an uncaught exception occurs or `print_fatal` is used.)

  - Add `pdb` command to drop to a shell.

  - Several new paths, `DATA_PATH`, `STATE_PATH` and XDG ones: The XDG basedir compliant paths.

  - Default DNS nameservers and hostname of generated host images can now be changed on command line.

  - Default exit code can now be configured as well (on command line as: `--default-exit_code`.)

  - Config file! Snr can now be configured with its config file inside `~/.config/snr/main.conf` (technically: `$XDG_CONFIG_HOME/snr/main.conf`)

  - `BlockInfo` a more pythonic way of handling block information.

  - Now you can set value of some flags using environment variables.

  - The init routine now has demands agreement to a disclaimer before continuing.

  - Fix a ton of type-annotation mistakes.

- Improvements:

  - Get rid of the confirmation on initialization.

  - `run_executable` (now `misc/run_executable`) payload searches for the executable better.

  - `clear` command now does a better job at clearing the screen.

  - Snr no longer requires root permissions (almost, with the exception of two parts.)

  - Snr no longer depends on `dosfstools`, `gdisk` and `parted`. (It uses the ones in the host image.)

  - `run_command` (now `misc/run_command`) now runs accepts several commands, and also writes them into a bash file to ensure better function.

  - `unix_user_management` is now called `tampering/user_management` will soon have windows support as well.

  - Merge `bios_disk_encryption` with `uefi_disk_encryption` (as `tampering/disk_encryption`.)

  - Use factory pattern for program wrappers.

  - Significantly improved CI/CD.

  - Fully comply with XDG basedir specifications.

  - Running snr from source code shows more information about the state of the version.
    Now, running from the source code shows git commit hash and whatever it was modified or not.
    Examples for the version (`snr.version.__version__`)
    Not running from source code:

    ```txt
    1.0.0
    ```

    Running from source code, no code modification:

    ```txt
    1.0.0-gitXXXXXXX
    ```

    Running from source code, with untracked or uncommitted modifications:

    ```txt
    1.0.0-gitXXXXXXX*
    ```

  - Convert docstrings to Google style

  - Stop the wrong process of importing everything in the library with an underscore (importing `os` as `_os`) and instead prefer google's method. Import only the modules so everything is well known
  (instead of `from snr.core.util.common_utils import print_ok as _print_ok` use `from snr.core.util import common_utils`.)

- Breaking changes:

  - Migrate from Debian stable rootfs to Ubuntu LTS rootfs.

  - Merge `libsnr` and `snr_payloads` with snr (`libsnr.core` is now `snr.core.core`.)

  - Remove the `Table` class and use the `rich.table` project instead.

  - Payloads are no longer one single python file but a package.

  - Move away from pure old pip to use poetry.

  - Use `click` and `prompt-toolkit` instead of a homegrown solution for commands.

  - Now payloads are categorized by what they do.

  - Remove the payload set update mechanism and library version check as they are no longer needed.
  
  - AtFormatter now looks for `"@variable@"` instead of `@variable@` to ensure that the code is still correct by syntax if the intended type is not string.

  - Version string is now in format: `snr {snr_version} {dependency}/{dependency_version}...`.

  - Using `--verbose` and `--quiet` now simply cancel each other out instead of whichever that came last.

  - Context is no longer a dictionary but a whole class by itself.

  - `ChrootProgramWrapper` no longer exists, use the factory at `core.util.chroot_program_wrapper`.

  - `options.prompt` no longer exists, instead use `snr.cli.interactive_shell.prompt`.

  - `misc/run_command` now uses `COMMANDS`, and not `COMMAND`.

  - Now host rootfs image is inside the share directory, not the cache directory.

  - Storage utilities in `snr.core.payload.storage` now function with `BlockInfo`.

  - Remove `version.DEVELOPMENT`.

  - Remove snr development options (and subsequently `dev_snr` script).

  - `-V` is no longer a recognized flag, and `-v` is now for `--verbose`, to get version, use `--version`.

  - `-c` and `--code` flag no longer exist, snr no longer supports running snr commands non-interactively (which also means `options.interactive` no longer exists) And we would advise against running snr non-interactively using pipes.

  - `snr.core.util.programs` is no longer a directory of program wrappers, but a file with all the program wrappers.

snr/core/core/options.py
"""
User-passed options and shared variables
"""
import configparser
import os.path
import socket
import types
from typing import Mapping, Optional, TypedDict

from snr.core.core import common_paths

__all__ = (
    "debug", "verbose",
    "quiet", "initializing",
    "default_exit_code", "arch",
    "PROMPT_UNLOADED", "PROMPT_LOADED_FORMAT",
    "payload_path", "payload_module",
    "MINIMUM_TARGET_SIZE", "default_hostname",
    "default_primary_nameserver", "default_secondary_nameserver",
    "default_user_agent"
)

_ValidConfigValueType = type[str | int | bool]


class _TypeSchemaDictRequired(TypedDict):
    type: _ValidConfigValueType


class _TypeSchemaDict(_TypeSchemaDictRequired, TypedDict, total=False):
    alphanum_only: Optional[bool]
    ipaddress_only: Optional[bool]


_CONFIG_SCHEMA: Mapping[str, _TypeSchemaDict] = {
    "debug": {
        "type": bool,
    },
    "verbose": {
        "type": bool,
    },
    "quiet": {
        "type": bool,
    },
    "arch": {
        "type": str,
    },
    "default_exit_code": {
        "type": int,
    },
    "default_hostname": {
        "type": str,
        "alphanum_only": True
    },
    "default_primary_nameserver": {
        "type": str,
        "ipaddress_only": True
    },
    "default_secondary_nameserver": {
        "type": str,
        "ipaddress_only": True
    },
    "default_user_agent": {
        "type": str
    }
}

# Whatever snr is run in verbose mode or not
verbose: bool = False

# Whatever or not snr is run in quiet mode
quiet: bool = False

# Whatever or not snr is initializing
initializing: bool = False

# Snr's default exit code
default_exit_code: int = 0

# System architecture
arch: str

# Prompt when no payload is loaded
PROMPT_UNLOADED = "Snr> "

# Format for the prompt when a payload is loaded
PROMPT_LOADED_FORMAT = "Snr ({0})> "

# Path to the currently loaded payload
payload_path: str = ""

# Python module of the loaded payload or None if no payload is loaded
payload_module: types.ModuleType | None = None

# Minimum size all target mediums must at least be
MINIMUM_TARGET_SIZE = 2000 * 1024 * 1024

# Default hostname of generated images
default_hostname: str = "snr"

# Default primary dns of generated images
default_primary_nameserver: str = "1.1.1.1"

# Default secondary dns of generated images
default_secondary_nameserver: str = "1.0.0.1"

# Default user agent
default_user_agent: str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/113.0"

# If catching an exception, drop to a debug shell
debug: bool = False


def _validate_and_apply_str(option: str, value: str) -> None:
    if _CONFIG_SCHEMA[option].get("alphanum_only", False):
        if not value.isalnum():
            print(f"--> Configuration value of '{option}'"
                  "is not a valid alphanumeric string")
            return
        if _CONFIG_SCHEMA[option].get("ipaddress_only", False):
            try:
                socket.inet_pton(socket.AF_INET, value)
            except OSError:
                print(f"--> Configuration value of '{option}'"
                      "is not a valid ip address")
                return
    globals()[option] = value


def _validate_and_apply_int(option: str, value: str) -> None:
    try:
        int(value, base=10)
    except ValueError:
        print(f"--> Configuration value of '{option}' is not a valid integer")
        return
    globals()[option] = int(value, base=10)


def _validate_and_apply_bool(option: str, value: str) -> None:
    if value.lower() in ("y", "on", "yes"):
        globals()[option] = True
    elif value.lower() in ("n", "off", "no"):
        globals()[option] = False
    else:
        print(f"--> Configuration value of '{option}' is not a valid boolean")


def _validate_and_apply_config(config_parser: configparser.ConfigParser) -> None:
    global verbose  # pylint: disable=global-statement
    global quiet  # pylint: disable=global-statement
    for option, value in config_parser.items("main"):
        if option not in _CONFIG_SCHEMA:
            print(f"--> Configuration key '{option}' is not recognized")
            continue
        if _CONFIG_SCHEMA[option]["type"] is bool:
            _validate_and_apply_bool(option, value)
        if _CONFIG_SCHEMA[option]["type"] is int:
            _validate_and_apply_int(option, value)
        if _CONFIG_SCHEMA[option]["type"] is str:
            _validate_and_apply_str(option, value)
        # Just like on command line, passing both cancel each other out
        if quiet and verbose:
            print("--> Configuration specifies both quiet and verbose as true,"
                  "they will cancel each other out")
            quiet, verbose = False, False


def _parse_config() -> None:
    if not hasattr(_parse_config, "has_already_done"):
        config_parser = configparser.ConfigParser()
        config_parser.read_dict(
            {
                "main": {
                    "verbose": verbose,
                    "quiet": quiet,
                    "default_exit_code": default_exit_code,
                    "default_hostname": default_hostname,
                    "default_primary_nameserver": default_primary_nameserver,
                    "default_secondary_nameserver": default_secondary_nameserver,
                    "default_user_agent": default_user_agent
                }
            }
        )
        config_parser.read(os.path.join(
            common_paths.CONFIG_PATH, "main.conf"))
        _validate_and_apply_config(config_parser)
